baseapp/baseapp.go

```diff
@@ -534,10 +516,32 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		ctx = ctx.WithMultiStore(msCache)
 	}
 
-	result = handler(ctx, msg)
+	// No need to run Msg checks when running CheckTx(), since CheckTx() only needs to validate that fees can be paid
+	// Then when DeliverTx() is called, this will run, preventing us from doing the same checks twice
+	if mode != runTxModeCheck {
+		var msg = tx.GetMsg()
+		if msg == nil {
+			return sdk.ErrInternal("Tx.GetMsg() returned nil").Result()
+		}
+
+		// Validate the Msg.
+		err := msg.ValidateBasic()
+		if err != nil {
+			err = err.WithDefaultCodespace(sdk.CodespaceRoot)
+			return err.Result()
+		}
+
+		// Match route.
+		msgType := msg.Type()
+		handler := app.router.Route(msgType)
+		if handler == nil {
+			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
+		}
+		result = handler(ctx, msg)
 
-	// Set gas utilized
-	result.GasUsed = ctx.GasMeter().GasConsumed()
+		// Set gas utilized
+		result.GasUsed = ctx.GasMeter().GasConsumed()
+	}
```

Comment:
👍 yup this section looks good - you have simply consolidated logic which does not need to be run during check-tx